pl.c

```diff
@@ -653,10 +653,9 @@ struct track_info *pl_play_selected_row(void)
 		if (shuffle && !pl_empty(pl_visible)) {
 			struct shuffle_track *st = shuffle_list_get_next(&pl_visible->shuffle_root, NULL, pl_dummy_filter);
 			struct simple_track *track = &st->simple_track;
-			rv = pl_play_track(pl_visible, track);
+			rv = pl_play_track(pl_visible, track, true);
```

Comment:
It makes the intended behaviour a clearer, reduces code duplication, and it prevents calling `pl_select_playing_track` twice, which while harmless, is incorrect.

Reply:
I don't really understand. Removing this parameter again wouldn't add any code duplication, and the bug would still be fixed. The bug is that we added a manual `pl_select_playing_track()` _before_ `pl_play_selected_track()`, so removing the `pl_select_playing_track()` is what counts. We probably do disagree on "intended behaviour"; I think we should leave the first track selected unless `follow` is set, in which case `pl_play_track()` will handle it for us.

At the end of the day ignoring the `follow` setting is a minor issue, so overall +1 if you're not convinced.

Reply:
> probably do disagree on "intended behaviour

At first I was thinking the same thing, but after considering it more I realized why @edclements probably chose to do it this way. IMHO, it makes sense if you consider it from a usability perspective. From my understanding, the idea of `follow` is to make it easy to locate the current track being played, whether it is due to an action by you or by cmus. A user would disable it if they wanted to have sole control over the location of the current selection. In contrast, pl_play_selected row is a user action, and most importantly, a UI action. It happens when a user presses enter on, or clicks the selected playlist with the cursor in the list on the left side. This will already switch the second pane to the first track in the list (even if another track was already selected in the same playlist), making it easy to see what actually happened when you selected it while also making it easier choose another.

With only the fix and nothing more, it would jump to the middle of the shuffle list, but still jump the unfocused selection in the second pane to the first track, which is counter-intuitive. By forcing `follow` for this, it makes this behaviour consistent with regards to what seems to be the intended purpose and semantics of the two-paned UI and the `follow` option.

Nevertheless, I see where you are coming from. In the library view, it's not like it skips to a random track when you select an album/artist in the left even if shuffle is enabled. An alternative fix would be to rotate the shuffle list to bring the first track in the playlist to the beginning of the shuffle list. This would not change the behaviour in any other situation, and could arguably be more logical.

Ultimately, it comes down to whether the first track, for the purposes of the UI, should be defined as the first track to be played in the current state (I guess this also raises the question of whether the shuffle list implementation is an implementation detail or not), or if it should be defined as the track at the top of the list.

@flyingmutant, do you have any opinion on this?

Reply:
IMO current behaviour is OK. I find it somewhat irritating that the shuffling is the same every time I press enter on the same playlist, but that is definitely something for another pull request.

Reply:
I'm okay with it too then.

Reply:
By the way, the lack of reshuffle is already covered as part of #1045. :slightly_smiling_face: